contents/java/JavaReflections.md

```diff
@@ -1,173 +1,160 @@
 
-        if (age > 20) {
-            setHasWool();
-        }
-    }
+#### Updating private fields
 
-    private int getPrivateAge() {
-        // Since class is a reserved keyword, we use clazz in variable names
-        Class<Animal> animalClazz = Animal.class;
+Suppose that you need to write tests for `Sheep`. As part of setting up the test, you need to create a sheep with age = 20. Suppose that the age of the sheep is updated automatically as time passes, whereby the age increases by 1 after every minute. A naive way of creating a sheep with age = 20 is to simply wait for 20minutes before performing the test:
 
-        try {
-            Field f = animalClazz.getDeclaredField("age");
-            f.setAccessible(true);
-            return (int) f.get(this);
-        } catch (Exception e) {
-            e.printStackTrace();
-        }
+```java
+@Test
+public void foo() {
+  Sheep sheep = new Sheep();
+  TimeUnit.MINUTES.sleep(20);
+  // perform test
+}
+```
 
-        return 0;
-    }
+Alternatively, we can set a value to the private field using Reflection:
 
-    // Other methods here...
+```java
+@Test
+public void foo() {
+  Sheep sheep = new Sheep();
+  Field field = Sheep.class.getDeclaredField("age");
+  field.setAccessible(true);
+  field.set(sheep, 20);
+  // perform test
 }
 ```
 
-And there you have it! What `Sheep` is really doing is to examine itself at runtime in order to obtain the `age` field
-inherited from `Animal`. If it cannot, it assumes its `age` is zero. This technique can be used in test cases to
-access private fields and methods in the class under test without modifying the behaviour of the class itself.
+#### Testing private methods
 
-You may notice that the `Sheep#getPrivateAge()` method sets the age `Field` object to be accessible and might wonder
-the implications. Fret not! The `Field#getDeclaredField()` actually returns a new `Field` instance - so you're just
-setting that particular local `Field` instance to be accessible, not the actual `age` field itself. You can read more
-about it in this [StackOverflow question](http://stackoverflow.com/questions/10638826/java-reflection-impact-of-setaccessibletrue).
+Suppose you want to test `getAge()`. However, you are only able to indirectly do so by testing `isProducingWool()`. This is not good as we are not able to directly verify the age of a sheep. However, with the help of Reflection, we can now test private methods.
+
+```java
+@Test
+public void foo() {
+  Sheep sheep = new Sheep();
+  Method method = Sheep.class.getDeclaredMethod("getAge");
+  method.setAccessible(true);
+  int age = (int) method.invoke(sheep);
```

Comment:
Maybe we can explain more about `method.invoke(sheep)`, if the method require parameters `int` then we also need to supply the arguments here or change the example here to take and supply some parameters.

Also, should need the `try` and `catch` block or `throws Exception`.

Reply:
> Maybe we can explain more about method.invoke(sheep), if the method require parameters int then we also need to supply the arguments here or change the example here to take and supply some parameters.

I think there's no need to cos this is meant to be a simple guide? :)